csharp/src/Apache.Arrow/Ipc/DictionaryMemo.cs

```diff
@@ -0,0 +1,113 @@
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Apache.Arrow.Types;
+
+namespace Apache.Arrow.Ipc
+{
+    class DictionaryMemo
+    {
+        private Dictionary<long, IArrowArray> _idToDictionary;
+        private Dictionary<long, IArrowType> _idToValueType;
+        private Dictionary<Field, long> _fieldToId;
+
+        public DictionaryMemo()
+        {
+            _idToDictionary = new Dictionary<long, IArrowArray>();
+            _idToValueType = new Dictionary<long, IArrowType>();
+            _fieldToId = new Dictionary<Field, long>();
+        }
+
+        public IArrowType GetDictionaryType(long id)
+        {
+            if (!_idToValueType.TryGetValue(id, out IArrowType type))
+            {
+                throw new ArgumentException($"Dictionary with id {id} not found");
+            }
+            return type;
+        }
+
+        public IArrowArray GetDictionary(long id)
+        {
+            if (!_idToDictionary.TryGetValue(id, out IArrowArray dictionary))
+            {
+                throw new ArgumentException($"Dictionary with id {id} not found");
+            }
+            return dictionary;
+        }
+
+        public void AddField(long id, Field field)
+        {
+            if (_fieldToId.ContainsKey(field))
+            {
+                throw new ArgumentException($"Field {field.Name} is already in Memo");
+            }
+
+            if (field.DataType.TypeId != ArrowTypeId.Dictionary)
+            {
+                throw new ArgumentException($"Field type is not DictionaryType: Name={field.Name}, {field.DataType.Name}");
+            }
+
+            IArrowType valueType = ((DictionaryType)field.DataType).ValueType;
+
+            if (_idToValueType.TryGetValue(id, out IArrowType valueTypeInDic))
+            {
+                if (valueType != valueTypeInDic)
+                {
+                    throw new ArgumentException($"Field type {field.DataType.Name} does not match the existing type {valueTypeInDic})");
+                }
+            }
+
+            _fieldToId.Add(field, id);
+            _idToValueType.Add(id, valueType);
+        }
+
+        public long GetId(Field field)
+        {
+            if (!_fieldToId.TryGetValue(field, out long id))
+            {
+                throw new ArgumentException($"Field with name {field.Name} not found");
+            }
+            return id;
+        }
+
+        public long GetOrAssignId(Field field)
+        {
+            if (!_fieldToId.TryGetValue(field, out long id))
+            {
+                id = _fieldToId.Count + 1;
+                AddField(id, field);
+            }
+            return id;
+        }
+
+        public void AddOrReplaceDictionary(long id, IArrowArray dictionary)
+        {
+            _idToDictionary[id] = dictionary;
+        }
+
+        public void AddDictionaryDelta(long id, IArrowArray dictionary)
```

Comment:
Can this be deleted until it is supported? It is dead code.

Reply:
Fixed it.

Reply:
This hasn't been deleted (yet)

Reply:
Sorry, I didn't commit the change.